pkg/rt/manager.go

```diff
@@ -67,33 +75,47 @@ func (m *Manager) Refresh() {
 	m.mtx.Lock()
 	defer m.mtx.Unlock()
 
-	nextgen := map[string]interrupt{}
-	for _, id := range m.ids.ServiceIDs() {
-		if irq, ok := m.managed[id]; ok {
-			level.Debug(m.logger).Log("service_id", id, "subscriber", "maintain")
-			nextgen[id] = irq // move
-			delete(m.managed, id)
-		} else {
-			level.Info(m.logger).Log("service_id", id, "subscriber", "create")
-			nextgen[id] = m.spawn(id)
+	nextgen := map[subscriberKey]interrupt{}
+	for _, product := range api.Products {
+		if m.productCache.HasAccess(product) {
+			for _, id := range m.ids.ServiceIDs() {
+				key := subscriberKey{serviceID: id, product: product}
+
+				if irq, ok := m.managed[key]; ok {
+					level.Debug(m.logger).Log("service_id", id, "type", product, "subscriber", "maintain")
+					nextgen[key] = irq // move
+					delete(m.managed, key)
+				} else {
+					level.Info(m.logger).Log("service_id", id, "type", product, "subscriber", "create")
+					nextgen[key] = m.spawn(id, product)
+				}
+			}
 		}
-	}
 
-	for _, id := range m.managedIDsWithLock() {
-		level.Info(m.logger).Log("service_id", id, "subscriber", "stop")
-		irq := m.managed[id]
-		irq.cancel()
-		err := <-irq.done
-		delete(m.managed, id)
-		level.Debug(m.logger).Log("service_id", id, "interrupt", err)
-	}
+		for _, key := range m.managedKeysWithLock() {
```

Comment:
Should we iterate over `m.managed` directly here instead of using `m.managedKeysWithLock()`? We're already holding the lock. ` managedKeysWithLock()` gives us a slice sorted by service id, but we don't need a slice or the results to be sorted.  We just need to iterate over the subscriber keys that are still left in `m.managed`.